### ISSUE TITLE
bounds checking on nonlinear expression indices

### ISSUE THREAD

**Issue by rgiordan** [JuliaOpt/JuMP.jl#414]
I've noticed some counter-intuitive behavior when accessing nonlinear expressions using single indices.  It seems that you are allowed to set indexed nonlinear expressions one at a time and sum them, but that later assignments overwrite the whole array.  Here's an example:
```
m = Model()
@defVar(m, baz)
@defNLExpr(foo[1], 5 * baz)
@defNLExpr(foo[2], -3 * baz)
@defNLExpr(bar, sum{foo[i], i=1:2})
setValue(baz, 2)
ReverseDiffSparse.getvalue(bar, m.colVal) # Expect 4, get -12
```
You get different behavior with ordinary expressions:
```
m = Model()
@defVar(m, baz)
@defExpr(foo[1], 5 * baz)
@defExpr(foo[2], -3 * baz)

julia> foo[1]
ERROR: key not found: (1,)
 in getindex at /home/rgiordan/.julia/v0.3/JuMP/src/JuMPDict.jl:30

julia> foo[2]
-3 baz
```
It's not hard to avoid using this kind of expression, though it would be convenient.  For example, this works:
```
m = Model()
@defVar(m, baz)
@defNLExpr(foo[s=1:2], (s == 1) * 5 * baz + (s == 2) * (-3 * baz))
@defNLExpr(bar, sum{foo[i], i=1:2})
setValue(baz, 2)
ReverseDiffSparse.getvalue(bar, m.colVal) 
```
However, I think the single indices should at least fail with an error rather than allowing you to quietly overwrite the rest of the array.

**Comment by joehuchette** [comment]
The reason it's behaving this way is that the ``@defNLExpr`` macro expands to something like
```jl
foo = ...
```
so multiple calls will just overwrite the local foo variable, rather than append as you expected. A workaround could be something like
```jl
foo = Array(Any, 2)
foo[1] = @defNLExpr(...)
foo[2] = @defNLExpr(...)
```
We've toyed with the idea of erroring instead as you suggest, but it's a surprisingly subtle issue. Julia only offers a rather rudimentary way to check if a variable is defined in any enclosing scope, so if you try to do something like e.g. ``isdefined(:foo)`` it will return ``true`` if there is any variable ``foo`` in an enclosing scope. This isn't what you want if, for instance, you're writing a model in a function, or if you try to use an identifier that's exported from Base as a variable name.

**Comment by joehuchette** [comment]
The reason it's behaving this way is that the ``@defNLExpr`` macro expands to something like
```jl
foo = ...
```
so multiple calls will just overwrite the local foo variable, rather than append as you expected. A workaround could be something like
```jl
foo = Array(Any, 2)
foo[1] = @defNLExpr(...)
foo[2] = @defNLExpr(...)
```
We've toyed with the idea of erroring instead as you suggest, but it's a surprisingly subtle issue. Julia only offers a rather rudimentary way to check if a variable is defined in any enclosing scope, so if you try to do something like e.g. ``isdefined(:foo)`` it will return ``true`` if there is any variable ``foo`` in an enclosing scope. This isn't what you want if, for instance, you're writing a model in a function, or if you try to use an identifier that's exported from Base as a variable name.

**Comment by mlubin** [comment]
Actually you can't manually create an array of nonlinear expression objects because they're a bit funky. There's no error in the nonlinear case because we don't that indices are valid (but we should). Anyway the issue is the same as with
```
@defVar(m, x[1])
@defVar(m, x[2])
```
which is equivalent to
```
x = # array of variables indexed by [1]
x = # array of variables indexed by [2]
```
since it's just an assignment we have to follow julia's semantics here and the original ``x`` is lost. There isn't really a sane way to check if a variable with the same name already exists in the *local* scope and override Julia's rules for assigning variables.

This is almost a dup of #105, but leaving open since we need to do more checks on if indices to nonlinear expressions are valid.

**Comment by rgiordan** [comment]
Cool, that makes sense.  Thanks as always for the quick reply.

The workaround of having an array of expressions is a neat idea, but it doesn't seem to work as-is.  The big problem is that taking the sum of an array of NLExpr objects doesn''t work.  An example follows.  This isn't a big deal unless you do want to recommend the above workaround.
```
m = Model()
@defVar(m, bar)

foo = Array(Any, 5, 10)
for i = 1:5
	for j = 1:10
		foo[i, j] = @defNLExpr(placeholder, i * bar + j)
	end
end
setValue(bar, 2)

# This works:
foo_mat = [ ReverseDiffSparse.getvalue(foo[i, j], m.colVal) for i=1:5, j=1:10]

# The result of this is a parametric expression, not a value:
@defNLExpr(foo_element, foo[1, 1])
ReverseDiffSparse.getvalue(foo_element, m.colVal)

# The result of this is an error:
@defNLExpr(foo_sum, sum{foo[i, j], i=1:5, j=1:10})
ReverseDiffSparse.getvalue(foo_sum, m.colVal)
# ERROR: `+` has no method matching +(::Float64, ::ParametricExpression{0})
#  in _EXPRVAL_ at /home/rgiordan/.julia/v0.3/ReverseDiffSparse/src/revmode.jl:678
#  in getvalue at /home/rgiordan/.julia/v0.3/ReverseDiffSparse/src/revmode.jl:696
```

**Comment by mlubin** [comment]
Arrays of expressions just won't work at the moment. The only syntax to declare a group of expressions is through ``@defNLExpr``.
If you have small number of different cases, you can do
```
@defNLExpr(foo[s=1:2], (s == 1) * 5 * baz + (s == 2) * (-3 * baz))
```
as you suggested. It's a bit more efficient for nonobvious reasons to do
```
@defNLExpr(foo[s=1:2], sum{5 * baz; s== 1} + sum{-3 * baz; s== 2})
```
but I wouldn't worry about the subtleties unless you find that derivative evaluation is unexpectedly slow.

**Comment by mlubin** [comment]
Will be fixed by #638, nonlinear expressions will use the same container infrastructure as the rest of JuMP and will have bounds checks
